### ISSUE TITLE
[BUG] Task state with CuPy array

### ISSUE THREAD

**Issue by matthieubulte** [dask/distributed#2837]
Hi!

This issue has started as an issue in the `dask_cuda` project, but it seems to be more adequate here. The original issue can be found here: https://github.com/rapidsai/dask-cuda/issues/93

I am currently working on implementing some statistical estimators using `dask`, `distributed`, `cupy` and `dask_cuda` for multi-GPU support.

My issue all started with the observation I am doing in the referred issue: the tasks screen of the dashboard indicates very high and unintended `transfer-*`costs when performing some simple operations on the GPU. Please see the initial issue to have a screenshot of the task screen.

After some exploration of the execution via some custom `SchedulerPlugin` (this is a great feature!), I realized that these very high transfer cost (>500ms) didn't make sense. They were associated to `gather` actions on tasks gathering chunks smaller than 1kb. See:
![image](https://user-images.githubusercontent.com/6624306/61159113-400f4580-a4fb-11e9-9b11-7c19307c6f33.png)

After some more debugging, I realized that the reason for that is that `cupy` operations are asynchronous. You can find an explanation and example in the original issue. However, since `distributed` is not aware of this, the library is mislead in measuring very low `compute-*` times and very high `transfer-*` time since the serialization requires a synchronous copying from the GPU and the CPU, which blocks until the computation is over and the data is copied.

I know that `distributed` and `cupy` are two very separate libraries, but this clearly seems to be a bug, or at least a misleading situation for people trying to combine `cupy` and `distributed`.

So, I would be very happy to work on a fix for this, but I'm not sure if you would be interested in adding a custom support for `cupy`, or if you would prefer to have (or already have) some generic solution for handling such situations.

NOTE: this is somewhat separated, but the attached screenshot was constructed via a mix of the graph computed by dask and the execution trace obtained with a `SchedulerPlugin`. Would a cleaner / nicer version of this be interesting to a more general audience?

**Comment by pentschev** [comment]
> So, I would be very happy to work on a fix for this (on the cupy side, it would be something like adding `cupy_array.device.synchronize()`, see https://docs-cupy.chainer.org/en/stable/reference/generated/cupy.cuda.Device.html#cupy.cuda.Device.synchronize ), but I'm not sure if you would be interested in adding a custom support for `cupy`

Should a specialization like this be done, (most likely) the correct place for that would be https://github.com/rapidsai/dask-cuda.

> or if you would prefer to have (or already have) some generic solution for handling such situations.

If there's a problem that can be generalized to distributed (e.g., this problem can be reproduced with some non-GPU library too) then it would probably be the case that we would attempt to have a generic solution in distributed itself.

Maybe @mrocklin can comment on this, and also if he's aware of any such cases that do not involve CuPy specifically.

**Comment by mrocklin** [comment]
> I would be very happy to work on a fix for this

I'd be very happy to see a fix for this, and would be even happier to see you work on it :)

> but I'm not sure if you would be interested in adding a custom support for cupy

Certainly not in the core scheduling algorithms.  They don't even have custom support for NumPy/Pandas.

> or if you would prefer to have (or already have) some generic solution for handling such situations.

There are other plugin mechanisms, like `WorkerPlugin`s, that could be modified to call something after every task transition within a worker.  Maybe you could extend things here with a custom plugin that handled cupy intricacies?

I fully acknowledge that this is a problem, but I also don't have any suggestions on how to solve it today.  I do encourage folks to search for a solution though, given the constraint that we won't insert specializations for particular data types within the core dask logic.

**Comment by matthieubulte** [comment]
Great, I will see what I can do.

These are the actions that I understand from your comment:
- Extend the `WorkerPlugin` API and `Worker` class to add a `post_execute` hook. I guess we could also add some `pre_execute` hook as well, this could be useful for finer resource management for instance;
- Call the `post_execute` hook at the right place in the task execution part of `Worker`. I still have to understand where this happens exactly;
- Create a plugin that would detect that the result of a task is a `cp.ndarray` or `cp.dataframe` and sync with the GPU if necessary.

Is that correct?

**Comment by mrocklin** [comment]
That is one way of going about things, yes.

Also, if all you want is to synchronize at every step then I suspect that
cupy has a mechanism to do this optionally already.  That might be
simpler.  Presumably there is a reason why they don't synchronize at every
step though.  I suspect that this might impact overall performance.

**Comment by pentschev** [comment]
> Also, if all you want is to synchronize at every step then I suspect that cupy has a mechanism to do this optionally already. That might be simpler. Presumably there is a reason why they don't synchronize at every step though. I suspect that this might impact overall performance.

You can let CuPy run synchronous, which means it will block during every memory copy or kernel execution, yes. But that's indeed going to badly hurt performance in most cases, so I'd advise against that and only ensure synchronization when really needed, such as when copying data back to CPU, for example.

**Comment by matthieubulte** [comment]
> Presumably there is a reason why they don't synchronize at every
step though.  
Well, I would guess because there is no use in doing so in the normal cupy setup running against one GPU. In the dask case where tasks are scheduled according to worker availability this is a different story I would say.

I think a (pretty bad) alternative is to replace the dask-managed computation by a normal cupy implementation in `map_blocks` with partitions of the size of the GPU memory and sync at the end of the computation on of each partition. That way, no overhead due to useless sync and the scheduler doesn't get confused. This would be quite sad however since it wouldn't give access to the optimizer and dynamic scheduler.

I'll give the global sync a try using the startup hook already available on the workers!

> But that's indeed going to badly hurt performance in most cases
That might be true. However to my understanding:
1. The computing time lost is only the idle time between the moment the worker "thinks" it's done with its current task and the next task coming from the scheduler.
2. This breaks the dynamic scheduling. The scheduler has basically no idea of the workers are busy until data has to be copied between workers which is the only point where we actually wait for a computation to be done.

Maybe I'm missing something, but since we're not copying any data when syncing, why would it be a perf problem to sync with the GPU?

> You can let CuPy run synchronous
Ah, I didn't find how to do this. I'll look again, thanks!

**Comment by mrocklin** [comment]
> I think a (pretty bad) alternative is to replace the dask-managed computation by a normal cupy implementation in map_blocks with partitions of the size of the GPU memory and sync at the end of the computation on of each partition. That way, no overhead due to useless sync and the scheduler doesn't get confused. This would be quite sad however since it wouldn't give access to the optimizer and dynamic scheduler.

Right, so if you had a plugin that called cupy's synchronize function after every task then you would probably get pretty close to this.  Dask array should do a fair amount of fusion of embarrassingly parallel operations today.

**Comment by matthieubulte** [comment]
Yes, that's exactly what I want: add the sync at the right place and let the optimizer decide where to fuze instead of doing it by hand!

Thank you for the feedback, I'll come back with a small PR :)

**Comment by pentschev** [comment]
I think this is something that should be benchmarked to verify how performance is affected. In general, we should have the least synchronization calls as possible, so it would be good if there's a configuration to enable/disable synchronization.

**Comment by matthieubulte** [comment]
> I think this is something that should be benchmarked to verify how performance is affected. 

I agree, I was also thinking of preparing a suite of performance tests to understand cupy with / wo sync and with different sync granularity. Then the same tests running on dask distributed with different optimization levels and maybe even the number of GPUs. 

I think what might also be interesting is to see how different will the scheduling be under these different regimes.

> In general, we should have the least synchronization calls as possible, so it would be good if there's a configuration to enable/disable synchronization.

Agree, I think the test will help understand this trade-off.

**Comment by pentschev** [comment]
Thanks for working on this @matthieubulte!

**Comment by matthieubulte** [comment]
Alright, I have a first version of the benchmark.

The associated code change is: https://github.com/matthieubulte/distributed/commit/f4280434cccb11856e8456b48d5470f96740c093

This code change is not clean, but it's enough for benchmarking.

The result of the benchmark is here:

https://gist.github.com/matthieubulte/db1420e03a4287f253bddecda349dbdf

Conclusion:
- On single GPU without using dask, this has no impact
- On single or multiple GPU using dask. this has a consistent impact of 15% regardless of the number of GPUs.

I'm not sure to understand why the impact is so large, and why it is constant across configurations. I will spend some more time on it in the next couple of days, but eventually it probably makes sense to close this, sadly.

**Comment by pentschev** [comment]
I see now, we're always using the default CUDA stream for everything, and all calls are synchronous, so probably the only synchronization step that matters is the last one.

Regardless of performance numbers, I'm more interested in knowing whether this change fixes the seemingly wrong assumption of Dask that memory transfer is taking most of the time, as stated in https://github.com/rapidsai/dask-cuda/issues/93, does it?

What I think is happening is the last operation in the pipeline is consuming that 15% difference in compute time, but I think the best way to have an overview of what happens is to look at the NVIDIA Visual Profiler to understand how execution develops over time.

**Comment by matthieubulte** [comment]
Hi, sorry that I couldn't take care of this earlier.

> Regardless of performance numbers, I'm more interested in knowing whether this change fixes the seemingly wrong assumption of Dask that memory transfer is taking most of the time, as stated in rapidsai/dask-cuda#93, does it?

I re-ran the experiment with and without the "sync" plugin and here are the two task graphs:
Without the plugin:
![Capture](https://user-images.githubusercontent.com/6624306/61642210-d59b9980-aca0-11e9-97f6-dce5b749cfcb.PNG)

With the plugin:
![Capture2](https://user-images.githubusercontent.com/6624306/61642231-ddf3d480-aca0-11e9-981e-b2251e062af4.PNG)

So, memory transfer seems to be taking some time, but it's not as bad as the sync-less profile indicates.

**Comment by mrocklin** [comment]
Thanks for sharing.  That's interesting to see.

**Comment by matthieubulte** [comment]
Is there any next step on this? Do you have any opinion?

**Comment by pentschev** [comment]
Sorry for missing the reply here @matthieubulte.

I think it's useful to have this change and allow it to be enabled if necessary, provided synchronization doesn't happen by default, since it will likely slowdown some applications (as seen in the 15% penalty in the tests above). This seems to be the case in the way changes were made to `Worker` already, so I'm fine keeping it this way.

@mrocklin do you have any further comments? Otherwise I would suggest @matthieubulte to open a PR with his changes and discuss there if the code should be modified in any way.

**Comment by mrocklin** [comment]
> @mrocklin do you have any further comments? 

Not really

> Otherwise I would suggest @matthieubulte to open a PR with his changes and discuss there if the code should be modified in any way.

My understanding is that the changes in this case were an external plugin that was cupy specific.  If so, then I would prefer to see this maintained outside of the dask codebase.  It would be great to have a copy of the code posted here so that other people could see an example of how to do this.  I don't think that it's the sort of thing we would include though. 

It may be that @matthieubulte needed to make core changes to the plugin design to make his plugin work though.  In that case it would be great to have those changes submitted.

(I'm guessing here at the approach, please ignore this response if it is making wrong assumptions)

**Comment by matthieubulte** [comment]
> @matthieubulte to open a PR with his changes and discuss there if the code should be modified in any way.

Will do, I'm in my exams period right now, so it will have to wait for a month.

> My understanding is that the changes in this case were an external plugin that was cupy specific. If so, then I would prefer to see this maintained outside of the dask codebase. It would be great to have a copy of the code posted here so that other people could see an example of how to do this. I don't think that it's the sort of thing we would include though.

I agree, I'll prepare this as well!
